### PR TITLE
Make StateMachineCtap2 respect runloop cancellations

### LINKED ISSUE #227: StateMachineCtap2 does not respect runloop cancellations
`StateMachineCtap2` does not pass the `alive` indicator down into `send_msg`, `send_cbor`, etc. The result is that `cancel` does not work when the device is in `HIDDevice::sendrecv`, e.g. when the device is waiting for a test of user presence.

For CTAP2 devices we should also be sending an explicit [CTAPHID_CANCEL](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#usb-hid-cancel) message.

### PR DESCRIPTION
**Pull request by jschanck**
Resolves #227

### CODE REVIEW

**Review by mozkeeler — APPROVED** [review]
LGTM!

**Review by msirringhaus — CHANGES_REQUESTED** [review]
Some minor things

File: src/transport/hid.rs [hunk 1]
```diff
@@ -84,15 +84,31 @@ where
...
+            // it's waiting for user presence. The keep_alive function is used to cancel the
+            // transaction.
+            if !keep_alive() {
+                break;
+            }
+        }
+        // If this is a CTAP2 device we can tell the authenticator to cancel the transaction on its
+        // side as well. There's nothing to do for U2F/CTAP1 devices.
+        if self.get_device_info().cap_flags.contains(Capability::CBOR) {
+            self.u2f_write(u8::from(HIDCmd::Cancel), &[])?;
```
[the remark this answers is not in the document]
Comment (jschanck) [reply src/transport/hid.rs:105]: The read is not necessary but it gives us a sensible return value in both cases.

It looks like `HIDDevice::cancel()` is only used in `DeviceSelector::cancel_all` right now. I don't think its use there is entirely correct, as it doesn't really make sense to (only) send the `CTAPHID_CANCEL` message when we might have U2F devices connected.

In fact, if I plug in a U2F-only device alongside a CTAP2-capable device and select the CTAP2 device, the U2F device continues to blink. So I think `HIDDevice::cancel()` needs to do more than send `CTAPHID_CANCEL`. If so I shouldn't use it in `HIDDevice::sendrecv()`. It should also probably be renamed.

[the remark this answers is not in the document]
Comment (msirringhaus) [reply src/transport/hid.rs:105]: The problem is: I don't know of a way to cancel an ongoing U2F-request. `CTAPHID_CANCEL` was introduced in ctap2.0. U2F only seems to have: 

> [4.1.1 U2FHID_MSG](https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-hid-protocol-v1.2-ps-20170411.html#u2fhid_msg)
>    [4.1.2 U2FHID_INIT](https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-hid-protocol-v1.2-ps-20170411.html#u2fhid_init)
>    [4.1.3 U2FHID_PING](https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-hid-protocol-v1.2-ps-20170411.html#u2fhid_ping)
>    [4.1.4 U2FHID_ERROR](https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-hid-protocol-v1.2-ps-20170411.html#u2fhid_error)
> 
> 4.2 Optional commands
> 
>    [4.2.1 U2FHID_WINK](https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-hid-protocol-v1.2-ps-20170411.html#u2fhid_wink)
>    [4.2.2 U2FHID_LOCK](https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-hid-protocol-v1.2-ps-20170411.html#u2fhid_lock)

Thats why currently, we let them blink and run into the timeout at some point. It's not ideal, but I don't know of any other way right now.

[the remark this answers is not in the document]
Comment (msirringhaus) [reply src/transport/hid.rs:105]: Ok, actually U2F operations are non-blocking. That means, we can just basically send "whatever" to the token to make it stop, like WINK or PING. 
I tried to figure out what Chromium is doing, but couldn't. The Cancel-function says, this only works with CTAP2 devices and is a no-op for CTAP1 devs. I _think_ they are implicitly sending a WINK command.

[the remark this answers is not in the document]
Comment (jschanck) [reply src/transport/hid.rs:105]: We don't need to send anything to CTAP1 devices. But we do need to exit out of the loop in `send_ctap1_cancellable` when the device selector calls `cancel_all`. The existing `HIDDevice::cancel` function won't work for that. I'll update the patch.

[the remark this answers is not in the document]
Comment (jschanck) [reply src/transport/hid.rs:121]: With the new version, an unselected CTAP1 device blinks 3 or 4 times after `DeviceSelector::cancel_all` is sent but then it stops blinking. I'm not sure whether we can do better, maybe the hardware just blinks a few times after any activity?

**Comment by jschanck** [comment]
Requesting review for the second patch. thanks!

**Review by mozkeeler — APPROVED** [review]
LGTM!

**Pull request merged**